query-planner-js/src/conditions.ts

```diff
@@ -0,0 +1,229 @@
+import {
+  assert,
+  isVariable,
+  OperationElement,
+  Selection,
+  SelectionSet,
+  Variable
+} from "@apollo/federation-internals";
+import { extractOperationConditionals } from "@apollo/query-graphs";
+import { ConditionNode, PlanNode } from ".";
+
+export type VariableCondition = {
+  variable: Variable,
+  negated: boolean,
+}
+
+export type Condition = VariableCondition | boolean;
+
+// The invariant maintained on this type are that if is an array of variable conditions, then:
+// 1. that array is not empty (it has at least one condition).
```

Comment:
Possibly interesting, you can enforce a non-empty array at the type level:
```ts
type NonEmptyArray<T> = [T, ...T[]]
```

Reply:
Interesting. I wonder how well that actually work on more complex example, but I guess it works ok here so included a commit that adds that.